File: app/src/main/java/com/keplersegg/myself/Room/Dao/TaskDao.java
package com.keplersegg.myself.Room.Dao;

import android.arch.persistence.room.Dao;
import android.arch.persistence.room.Delete;
import android.arch.persistence.room.Insert;
import android.arch.persistence.room.OnConflictStrategy;
import android.arch.persistence.room.Query;
import android.arch.persistence.room.Update;

import com.keplersegg.myself.Room.Entity.Task;

import java.util.List;

@Dao
public interface TaskDao {

    @Query("SELECT * FROM Task")
    List<Task> getAll();

    @Query("SELECT * FROM Task where Id = :id")
    Task get(int id);

    @Query("SELECT count(0) FROM Task where Label = :label")
    int getCountByLabel(String label);

    @Query("SELECT * FROM Task where Id <> :taskId and Label = :label")
    int getCountByLabelExcludeId(int taskId, String label);

    @Query("SELECT COUNT(0) from Task")
    int getCount();

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    void insert(Task task);

    @Update(onConflict = OnConflictStrategy.IGNORE)
    int update(Task task);

    @Delete
    void delete(Task task);

    @Query("DELETE FROM Task")
    void deleteAll();
}

File: app/src/main/java/com/keplersegg/myself/Room/Dao/TaskDao.kt
package com.keplersegg.myself.Room.Dao

import com.keplersegg.myself.Room.Entity.Task

import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import androidx.room.Update

@Dao
interface TaskDao {

    @get:Query("SELECT * FROM Task")
    val all: List<Task>

    @get:Query("SELECT case when count(0) = 0 then 1 else max(Id)+1 end FROM Task")
    val maxId: Int

    @get:Query("SELECT COUNT(0) from Task")
    val count: Int

    @Query("SELECT * FROM Task where Id = :id")
    operator fun get(id: Int): Task

    @Query("SELECT count(0) FROM Task where Label = :label")
    fun getCountByLabel(label: String): Int

    @Query("SELECT * FROM Task where Id <> :taskId and Label = :label")
    fun getCountByLabelExcludeId(taskId: Int, label: String): Int

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    fun insert(task: Task)

    @Update(onConflict = OnConflictStrategy.IGNORE)
    fun update(task: Task): Int

    @Delete
    fun delete(task: Task)

    @Query("DELETE FROM Task")
    fun deleteAll()
}
